File: shapes/traditional/src/main/scala/shapes/util/Observable.scala
package shapes
package util

import scala.collection.mutable.ListBuffer

class Observable[T](var init: T) {
  val handlers = new ListBuffer[T => Unit]
  def addObserver(handler: T => Unit) = handlers += handler
  def set(v: T) = {
    init = v
    handlers foreach { _(init) }
  }
  def get: T = init
}

object Observable {
  def apply[T](v: T) = new Observable[T](v)
}

File: chat/common/js/src/main/resources/util/observable.js
"use strict"

function Observable(init) {
  this._value = init
  this._handlers = []
}

Observable.prototype.addObserver = function(handler) {
  this._handlers.push(handler)
}

Observable.prototype.set = function(value) {
  this._value = value
  this._handlers.forEach(function(handler) { handler(value) })
}

Observable.prototype.get = function() {
  return this._value
}
